ruma-federation-api/src/serde/v1_pdu.rs

```diff
@@ -1,42 +1,48 @@
-//! A module to deserialize a RoomState struct from incorrectly specified v1
-//! send_join endpoint.
+//! A module to deserialize a membership struct from incorrectly specified v1
+//! membership endpoint.
 //!
 //! For more information, see this [GitHub issue][issue].
 //!
 //! [issue]: https://github.com/matrix-org/matrix-doc/issues/2541
 
-use std::fmt;
+use std::{fmt, marker::PhantomData};
 
 use serde::{
-    de::{Deserializer, Error, IgnoredAny, SeqAccess, Visitor},
-    ser::{SerializeSeq, Serializer},
+    de::{Deserialize, Deserializer, Error, IgnoredAny, SeqAccess, Visitor},
+    ser::{Serialize, SerializeSeq, Serializer},
 };
 
-use crate::membership::create_join_event::RoomState;
-
-pub fn serialize<S>(room_state: &RoomState, serializer: S) -> Result<S::Ok, S::Error>
+pub fn serialize<T, S>(membership_struct: &T, serializer: S) -> Result<S::Ok, S::Error>
 where
     S: Serializer,
+    T: Serialize,
 {
     let mut seq = serializer.serialize_seq(Some(2))?;
     seq.serialize_element(&200)?;
-    seq.serialize_element(room_state)?;
+    seq.serialize_element(membership_struct)?;
     seq.end()
 }
 
-pub fn deserialize<'de, D>(deserializer: D) -> Result<RoomState, D::Error>
+pub fn deserialize<'de, T, D>(deserializer: D) -> Result<T, D::Error>
 where
     D: Deserializer<'de>,
+    T: Deserialize<'de>,
 {
-    deserializer.deserialize_seq(RoomStateVisitor)
+    deserializer.deserialize_seq(MembershipStructVisitor { phantom: PhantomData })
+}
+
+struct MembershipStructVisitor<T> {
```

Comment:
```suggestion
struct PduVisitor<T> {
```

or sth. like that. Depends on what we end up as the name for the module.